## Problema 1

<img src="problema1.jpg">

1. Escribe un código para determinar, mediante simulaciones, la probabilidad de que, escogiendo un punto al azar dentro del cuadrado, éste se encuentre dentro del círculo. Calcula esta probabilidad bajo al menos 3 distintos números de iteraciones, de distinto orden de magnitud.
2. Analíticamente, ¿Cuál debería ser esta probabilidad?
3. De acuerdo con 2), el código que realizaste en 1) sirve para aproximar ¿Qué famosa constante? Modifica el número de iteraciones en tu código hasta dar correctamente los primeros cuatro decimales de esta constante. ¿Cuántas iteraciones fueron suficientes?


### Codigo

In [7]:
import numpy as np
import math

#### 1.-Variables iniciales

In [25]:
vAciertos= 0 
N = 100

#### 2.- Generar N puntos aleatorios (x,y)

In [26]:
x, y = np.random.uniform(-0.5, 0.5, size=(2, N))

#### 3.- Bucle For que recorre los puntos

In [27]:
for (a, b) in zip (x, y):
    p= math.sqrt(a**2+b**2)
    if (p<0.5):
        vAciertos +=1
print(vAciertos)

78


#### 4.- Mostrando resultados

In [28]:
print("Número de aciertos: ", vAciertos)
print("Número de fallos: ", N-vAciertos)
print("Valor aproximado de Pi:", (4* (vAciertos/N)))

Número de aciertos:  78
Número de fallos:  22
Valor aproximado de Pi: 3.12


### Mejorando 
Entre más se aumente el tamaño del experimento se llegue a una cantidad más aproximada de PI, usemos 10,000,0000 como la cantidad de experimentos.


In [31]:
vAciertos= 0 
N = 200000000
x, y = np.random.uniform(-0.5, 0.5, size=(2, N))
for (a, b) in zip (x, y):
    p= math.sqrt(a**2+b**2)
    if (p<0.5):
        vAciertos +=1
print("Número de aciertos: ", vAciertos)
print("Número de fallos: ", N-vAciertos)
print("Valor aproximado de Pi:", (4* (vAciertos/N)))

Número de aciertos:  157073480
Número de fallos:  42926520
Valor aproximado de Pi: 3.1414696


Este metodo se llama Monte Carlo y tiene distintas aplicaciones, para el calculo aproximadamente se necesitan 10 veces el número de iteraciones del número significativo anterior para obtener uno nuevo. 

Con 100,000,000 solo conseguí 3 para obtener 3.1415 se necesitaran alrededor de mil millones de iteraciones. 

# Problema 2

In [33]:
# 1.- Determinar la lista:
lista = [1,3,5,1,7,4,8,2,2,4]
N= len(lista)
print("La longitud de la lista es: ", N)

La longitud de la lista es:  10


In [34]:
# 2.- Crear listas vacias una para guardar los leidos y otra para guardar los repetidos
listaLeidos =[]
listaDuplicados=[]


In [36]:
# 3.- Leer cada componente de la lista
for v in lista: 
    #3.1 verificar que v no esté en la lista de Duplicados
    if not(v in listaDuplicados):
        #3.1.1 Dentro del if, verificar si esta en la de lista de leidos y si lo está agregar el valor a la lista de duplicados
        if (v in listaLeidos): 
            listaDuplicados.append(v)
        #3.1.2 De lo contrario, agregar a la lista de leidos
        else: 
            listaLeidos.append(v)
print(lista)
print(listaLeidos)
print(listaDuplicados)

[1, 3, 5, 1, 7, 4, 8, 2, 2, 4]
[1, 3, 5, 7, 4, 8, 2]
[1, 2, 4]


In [37]:
# 1.- Determinar la lista:
lista = [1,3,5,1,7,4,8,2,2,4,15, 18, 32, 15]
N= len(lista)
print("La longitud de la lista es: ", N)
# 2.- Crear listas vacias una para guardar los leidos y otra para guardar los repetidos
listaLeidos =[]
listaDuplicados=[]
# 3.- Leer cada componente de la lista
for v in lista: 
    #3.1 verificar que v no esté en la lista de Duplicados
    if not(v in listaDuplicados):
        #3.1.1 Dentro del if, verificar si esta en la de lista de leidos y si lo está agregar el valor a la lista de duplicados
        if (v in listaLeidos): 
            listaDuplicados.append(v)
        #3.1.2 De lo contrario, agregar a la lista de leidos
        else: 
            listaLeidos.append(v)
print(lista)
print(listaLeidos)
print(listaDuplicados)


La longitud de la lista es:  14
[1, 3, 5, 1, 7, 4, 8, 2, 2, 4, 15, 18, 32, 15]
[1, 3, 5, 7, 4, 8, 2, 15, 18, 32]
[1, 2, 4, 15]
